modules/l4quic/matcher.go

```diff
@@ -69,35 +71,57 @@ func (m *MatchQUIC) Match(cx *layer4.Connection) (bool, error) {
 		return false, nil
 	}
 
-	// Read one byte
-	n := 1
-	buf := make([]byte, n)
-	_, err := io.ReadAtLeast(cx, buf, n)
+	// Read at least one byte
+	buf := bufPool.Get().([]byte)[:QUICPacketBytesMax+1]
+	defer bufPool.Put(buf)
+	n, err := io.ReadAtLeast(cx, buf, 1)
+	buf = buf[:n]
 	if err != nil {
 		return false, err
 	}
 
 	// Ensure the second bit of the first byte is set, i.e. continue if
 	// github.com/quic-go/quic-go/internal/wire.IsPotentialQUICPacket(buf[0]).
-	qFirstByte := buf[0]
-	if qFirstByte&QUICMagicBitValue == 0 {
+	if buf[0]&QUICMagicBitValue == 0 {
 		return false, nil
 	}
 
 	// Ensure the first bit of the first byte is set, i.e. continue if
 	// github.com/quic-go/quic-go/internal/wire.IsLongHeaderPacket(buf[0]).
 	// Note: this behaviour may be changed in the future if there are packets
 	// that should be considered valid despite having the first bit unset.
-	if qFirstByte&QUICLongHeaderBitValue == 0 {
+	if buf[0]&QUICLongHeaderBitValue == 0 {
 		return false, nil
 	}
 
-	// Read the remaining bytes
-	buf = make([]byte, QUICPacketBytesMax+1)
-	buf[0] = qFirstByte
-	n, err = io.ReadAtLeast(cx, buf[1:], 1)
-	if err != nil || n < QUICPacketBytesMin-1 || n == QUICPacketBytesMax {
-		return false, err
+	// Check frame size lies within acceptable range
+	if n < QUICPacketBytesMin || n > QUICPacketBytesMax {
+		return false, nil
+	}
+
+	// Read remaining frames
+	var bufs [][]byte
+	bufs = append(bufs, buf)
+	defer func() {
+		// First buffer will be put when it's acquired
+		for _, b := range bufs[1:] {
+			bufPool.Put(b)
+		}
+	}()
+
+	for {
+		subBuf := bufPool.Get().([]byte)[:QUICPacketBytesMax+1]
```

Comment:
We read up to `QUICPacketBytesMax+1` bytes here, but never (a) check that n is within the relevant range; (b) check the first byte has relevant bits set so that we're sure it represents a valid QUIC request fragment.